File: 257.binary-tree-paths.cpp
/*
 * @lc app=leetcode id=257 lang=cpp
 *
 * [257] Binary Tree Paths
 *
 * https://leetcode.com/problems/binary-tree-paths/description/
 *
 * algorithms
 * Easy (49.21%)
 * Likes:    1516
 * Dislikes: 93
 * Total Accepted:    302.2K
 * Total Submissions: 601.1K
 * Testcase Example:  '[1,2,3,null,5]'
 *
 * Given a binary tree, return all root-to-leaf paths.
 * 
 * Note: A leaf is a node with no children.
 * 
 * Example:
 * 
 * 
 * Input:
 * 
 * ⁠  1
 * ⁠/   \
 * 2     3
 * ⁠\
 * ⁠ 5
 * 
 * Output: ["1->2->5", "1->3"]
 * 
 * Explanation: All root-to-leaf paths are: 1->2->5, 1->3
 * 
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    vector<string> binaryTreePaths(TreeNode *root)
    {
        if (root == NULL)
        {
            return vector<string>{};
        }

        vector<string> res;

        vector<string> l = binaryTreePaths(root->left);
        res.insert(res.end(), l.begin(), l.end());
        vector<string> r = binaryTreePaths(root->right);
        res.insert(res.end(), r.begin(), r.end());

        for (int i = 0; i < res.size(); i++)
        {
            res[i] = to_string(root->val) + "->" + res[i];
        }

        if (res.size() == 0)
        {
            res.push_back(to_string(root->val));
        }

        return res;
    }
};
// @lc code=end

File: 257.binary-tree-paths/257.binary-tree-paths.java
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.ListIterator;

import org.graalvm.util.CollectionsUtil;

/*
 * @lc app=leetcode id=257 lang=java
 *
 * [257] Binary Tree Paths
 *
 * https://leetcode.com/problems/binary-tree-paths/description/
 *
 * algorithms
 * Easy (56.27%)
 * Likes:    3133
 * Dislikes: 154
 * Total Accepted:    440.7K
 * Total Submissions: 783.2K
 * Testcase Example:  '[1,2,3,null,5]'
 *
 * Given the root of a binary tree, return all root-to-leaf paths in any
 * order.
 * 
 * A leaf is a node with no children.
 * 
 * 
 * Example 1:
 * 
 * 
 * Input: root = [1,2,3,null,5]
 * Output: ["1->2->5","1->3"]
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: root = [1]
 * Output: ["1"]
 * 
 * 
 * 
 * Constraints:
 * 
 * 
 * The number of nodes in the tree is in the range [1, 100].
 * -100 <= Node.val <= 100
 * 
 * 
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<String> binaryTreePaths(TreeNode root) {
        if (root == null) {
            return new ArrayList<>();
        }

        if (root.left == null && root.right == null) {
            return new ArrayList<>(Arrays.asList(String.valueOf(root.val)));
        }

        List<String> res = binaryTreePaths(root.left);
        res.addAll(binaryTreePaths(root.right));

        ListIterator<String> iter = res.listIterator();
        while (iter.hasNext()) {
            String s = iter.next();
            iter.set(String.valueOf(root.val) + "->" + s);
        }

        return res;
    }
}
// @lc code=end

